File: backend/src/main/java/com/mbi_re/airport_management/service/BookingService.java
package com.mbi_re.airport_management.service;

import com.mbi_re.airport_management.config.TenantContext;
import com.mbi_re.airport_management.dto.BookingDTO;
import com.mbi_re.airport_management.model.Booking;
import com.mbi_re.airport_management.repository.BookingRepository;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
public class BookingService {

    private final BookingRepository repository;

    public BookingService(BookingRepository repository) {
        this.repository = repository;
    }

    /**
     * Creates a new booking for the specified tenant.
     * Evicts the bookings cache for the tenant to maintain cache consistency.
     *
     * @param dto      the booking data transfer object containing booking details
     * @param tenantId the tenant identifier to associate the booking with
     * @return the created Booking entity
     */
    @CacheEvict(value = "bookings", key = "#tenantId + '_all'")
    public Booking createBooking(BookingDTO dto, String tenantId) {
        Booking booking = new Booking();
        booking.setFlightNumber(dto.getFlightNumber());
        booking.setPassengerName(dto.getPassengerName());
        booking.setSeatNumber(String.join(",", dto.getSeatNumber()));
        booking.setPassengerId(dto.getPassengerId());
        booking.setBookingTime(LocalDateTime.now());
        booking.setStatus("PAID");
        booking.setBookingId(generateBookingId());
        booking.setTenantId(tenantId);

        return repository.save(booking);
    }

    /**
     * Retrieves all bookings for the current tenant.
     * Results are cached for performance improvement.
     *
     * @return list of Booking entities associated with the current tenant
     */
    @Cacheable(value = "bookings", key = "T(com.mbi_re.airport_management.config.TenantContext).getTenantId() + '_all'")
    public List<Booking> getAllBookings() {
        String tenantId = TenantContext.getTenantId();
        return repository.findByTenantId(tenantId);
    }

    /**
     * Retrieves a booking by its ID for the current tenant.
     * Returns null if the booking is not found or does not belong to the tenant.
     *
     * @param id the booking ID
     * @return the Booking entity if found and tenant matches, else null
     */
    public Booking getBookingById(Long id) {
        String tenantId = TenantContext.getTenantId();
        return repository.findByIdAndTenantId(id, tenantId).orElse(null);
    }

    /**
     * Updates an existing booking for the current tenant by booking ID.
     * Evicts the bookings cache for the tenant after update.
     *
     * @param id  the booking ID to update
     * @param dto the booking data transfer object containing updated fields
     * @return the updated Booking entity, or null if not found or unauthorized
     */
    @CacheEvict(value = "bookings", key = "T(com.mbi_re.airport_management.config.TenantContext).getTenantId() + '_all'")
    public Booking updateBooking(Long id, BookingDTO dto) {
        String tenantId = TenantContext.getTenantId();
        Optional<Booking> optional = repository.findByIdAndTenantId(id, tenantId);

        return optional.map(b -> {
            b.setPassengerName(dto.getPassengerName());
            b.setPassengerId(dto.getPassengerId());
            b.setBookingId(dto.getBookingId());
            b.setFlightNumber(dto.getFlightNumber());
            b.setSeatNumber(dto.getSeatNumber());
            b.setStatus(dto.getStatus());
            return repository.save(b);
        }).orElse(null);
    }

    /**
     * Deletes a booking by ID scoped to the current tenant.
     * Evicts the bookings cache for the tenant upon successful deletion.
     * The operation is transactional.
     *
     * @param id the booking ID to delete
     */
    @CacheEvict(value = "bookings", key = "T(com.mbi_re.airport_management.config.TenantContext).getTenantId() + '_all'")
    @Transactional
    public void deleteBooking(Long id) {
        String tenantId = TenantContext.getTenantId();
        Optional<Booking> optional = repository.findByIdAndTenantId(id, tenantId);
        optional.ifPresent(repository::delete);
    }

    /**
     * Updates the check-in status of a booking for the current tenant.
     *
     * @param id        the booking ID
     * @param checkedIn boolean flag indicating if the passenger has checked in
     * @return Optional containing the updated Booking if found, else empty
     */
    public Optional<Booking> updateCheckInStatus(Long id, boolean checkedIn) {
        String tenantId = TenantContext.getTenantId();
        return repository.findByIdAndTenantId(id, tenantId).map(booking -> {
            booking.setCheckedIn(checkedIn);
            return repository.save(booking);
        });
    }

    /**
     * Generates a unique booking ID string.
     * The format is "BOOK-" followed by an 8-character uppercase UUID substring.
     *
     * @return the generated unique booking ID string
     */
    private String generateBookingId() {
        return "BOOK-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
    }

}

File: frontend/src/api/bookingService.js
// src/services/bookingService.js
import axios from 'axios';

const API_URL = 'http://localhost:8080/api/bookings';

/**
 * Create a new booking (authenticated users only).
 * @param {Object} bookingDTO - Booking data to be submitted.
 * @param {string} token - JWT token for Authorization header.
 * @returns {Promise} Axios response promise.
 */
export const createBooking = (bookingDTO, token) => {
  return axios.post(API_URL, bookingDTO, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

/**
 * Get all bookings (admin only).
 * @param {string} token - JWT token for Authorization header.
 * @returns {Promise} Axios response promise.
 */
export const getAllBookings = (token) => {
  return axios.get(API_URL, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

/**
 * Get booking by ID (admin only).
 * @param {number} id - Booking ID.
 * @param {string} token - JWT token for Authorization header.
 * @returns {Promise} Axios response promise.
 */
export const getBookingById = (id, token) => {
  return axios.get(`${API_URL}/${id}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

/**
 * Update a booking (admin only).
 * @param {number} id - Booking ID.
 * @param {Object} bookingDTO - Updated booking data.
 * @param {string} token - JWT token for Authorization header.
 * @returns {Promise} Axios response promise.
 */
export const updateBooking = (id, bookingDTO, token) => {
  return axios.put(`${API_URL}/${id}`, bookingDTO, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

/**
 * Delete a booking by ID (admin only).
 * @param {number} id - Booking ID.
 * @param {string} token - JWT token for Authorization header.
 * @returns {Promise} Axios response promise.
 */
export const deleteBooking = (id, token) => {
  return axios.delete(`${API_URL}/${id}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

/**
 * Update check-in status (authenticated users).
 * @param {number} id - Booking ID.
 * @param {boolean} checkedIn - Check-in status to set (true or false).
 * @param {string} token - JWT token for Authorization header.
 * @returns {Promise} Axios response promise.
 */
export const updateCheckInStatus = (id, checkedIn, token) => {
  return axios.put(`${API_URL}/${id}/checkin?checkedIn=${checkedIn}`, {}, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};